### PR TITLE
Improve FP docs

### PR DESCRIPTION
**Pull request by h0tk3y**
* Massive reordering of the sections for better logical structure: it's now *Types* -> *Instances* -> *Usages*
* Remove duplicate explanations and examples, including those redundant with the new structure, provide links where needed
* Move sections about `it` and destucturing to the lambdas chapter
* Fix inline functions mentioned as a performance tool, which they are not; Tell about control flow instead
* Better structure of function types explanation, including `suspend` function types and function types with receivers, add a note on associativity, mention typealiases 
* Use `fold` as a single example in *Higher-Order Functions* (instead of `lock` and `map`)
* New section on instantiating a function type, with function type interchangeability explanation
* New section on invoking a functional value
* Rephrase the section on function literals with receivers to make it more clear

<hr>

* Add a section for callable references in general, mentioning their connection with the function types
* Add notes on types of callable references
* Add a section on bound inner class ctor reference

<hr>

* Introduce a lot of runnable samples

Fixes issues KT-17736 and KT-23258

### CODE REVIEW

**Comment by h0tk3y** [comment]
* Reorder the sections for better logical structure: it's now *Types* -> *Instances* -> *Usages*
* Remove duplicate explanations and examples, including those redundant with the new structure, provide links where needed
* Move sections about `it` and desturturing to the lambdas chapter
* Fix inline functions mentioned as a performance tool, which they are not; Tell about control flow instead
* Better structure of function types explanation, including `suspend` function types and function types with receivers, add a note on associativity, mention typealiases 
* New section on instantiating a function type, with function type interchangeability explanation
* New section on invoking a functional value
* Rephrase the section on function literals with receivers to make it more clear

* Add a section for callable references in general, mentioning their connection with the function types
* Add notes on types of callable references
* Add a section on bound inner class ctor reference

* Introduce a lot of runnable samples

Fixes issues KT-17736 and KT-23258

File: pages/docs/reference/lambdas.md [hunk 1]
```diff
@@ -7,156 +7,220 @@ title: "Higher-Order Functions and Lambdas"
...
+be stored in variables and data structures, passed as arguments to and returned from other 
+[higher-order functions](#higher-order-functions), and operated with in any way that is possible for other 
+non-function values. 
+
+To facilitate this, Kotlin, as a statically typed programming language, uses a family of 
+types to represent functions and provides a set of specialized language constructs, such as [lambda expressions](#lambda-expressions-and-anonymous-functions). 
+
+The function types have special notation that corresponds to the signatures of the functions:
+
+* All function types have a parenthesized parameter types list and a return type: `(A, B) -> C` denotes a type that consists of
```
Comment (yole) [inline pages/docs/reference/lambdas.md]: "consists of functions" - I'm not sure that it's the right terminology. "Represents functions"?

File: pages/docs/reference/lambdas.md [hunk 2]
```diff
@@ -7,156 +7,220 @@ title: "Higher-Order Functions and Lambdas"
...
+[higher-order functions](#higher-order-functions), and operated with in any way that is possible for other 
+non-function values. 
+
+To facilitate this, Kotlin, as a statically typed programming language, uses a family of 
+types to represent functions and provides a set of specialized language constructs, such as [lambda expressions](#lambda-expressions-and-anonymous-functions). 
+
+The function types have special notation that corresponds to the signatures of the functions:
+
+* All function types have a parenthesized parameter types list and a return type: `(A, B) -> C` denotes a type that consists of 
+ functions taking two arguments of types `A` and `B` and returning a value of type `C`.
```
Comment (yole) [inline pages/docs/reference/lambdas.md]: Note explicitly that `Unit` as a return type may not be omitted?

File: pages/docs/reference/lambdas.md [hunk 3]
```diff
@@ -7,156 +7,220 @@ title: "Higher-Order Functions and Lambdas"
...
+types to represent functions and provides a set of specialized language constructs, such as [lambda expressions](#lambda-expressions-and-anonymous-functions). 
+
+The function types have special notation that corresponds to the signatures of the functions:
+
+* All function types have a parenthesized parameter types list and a return type: `(A, B) -> C` denotes a type that consists of 
+ functions taking two arguments of types `A` and `B` and returning a value of type `C`. 
+ The parameter types list may be empty, as in `() -> A`. 
+ 
+* Function types can optionally have an additional *receiver* type, which is specified before a dot in the notation:
+ the type `A.(B) -> C` consists of functions that can be called on a receiver object of `A` with a parameter of `B`.
```
Comment (yole) [inline pages/docs/reference/lambdas.md]: ...and return `C`? (for clarity)

File: pages/docs/reference/lambdas.md [hunk 4]
```diff
@@ -7,156 +7,220 @@ title: "Higher-Order Functions and Lambdas"
...
 ```
+</div>
 
-These conventions allow to write [LINQ-style](http://msdn.microsoft.com/en-us/library/bb308959.aspx) code:
+In Kotlin, there is a convention that if the last parameter of a function accepts a function, a lambda expression that is 
+passed as the corresponding argument can be placed outside the parentheses:
 
 ``` kotlin
-strings.filter { it.length == 5 }.sortedBy { it }.map { it.toUpperCase() }
+val producetion = items.fold(1) { acc, e -> acc * e }
```
Comment (yole) [inline pages/docs/reference/lambdas.md]: product

File: pages/docs/reference/lambdas.md [hunk 5]
```diff
@@ -7,156 +7,220 @@ title: "Higher-Order Functions and Lambdas"
...
+        val result = acc + i
+		println("result = $result")
+        result
+    })
+    
+    // Parameter types in a lambda are optional if they can be inferred:
+    val joinedToString = items.fold("Elements:", { acc, i -> acc + " " + i })
+    
+    // Function references can also be used for higher-order function calls:
+    val production = items.fold(1, Int::times)
```
Comment (yole) [inline pages/docs/reference/lambdas.md]: product

File: pages/docs/reference/reflection.md [hunk 6]
```diff
@@ -40,45 +40,79 @@ val widget: Widget = ...
...
+You obtain the reference to an exact class of an object, for instance `GoodWidget` or `BadWidget`, despite the type of the receiver expression (`Widget`).
 
-## Function References
+## Callable references
+
+References to functions, properties, and constructors, apart from introspecting the program structure, can 
+also be called or used as instances of [function types](lambdas.html#function-types).
+
+The common supertype for all callable references is [`KCallable<out R>`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.reflect/-k-callable/index.html), 
+where `R` is the return value type.
```
Comment (yole) [inline pages/docs/reference/reflection.md]: Clarify what "return value type" means for properties?

**Pull request merged**